File: app/src/main/java/com/a60n1/ejashkojme/fragment/ConversationFragment.java
package com.a60n1.ejashkojme.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.a60n1.ejashkojme.ChatActivity;
import com.a60n1.ejashkojme.R;
import com.a60n1.ejashkojme.models.Conversation;
import com.a60n1.ejashkojme.models.Message;
import com.a60n1.ejashkojme.viewholder.ConversationViewHolder;
import com.firebase.ui.database.FirebaseRecyclerAdapter;
import com.firebase.ui.database.FirebaseRecyclerOptions;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.squareup.picasso.Callback;
import com.squareup.picasso.NetworkPolicy;
import com.squareup.picasso.Picasso;

import java.util.Objects;

public class ConversationFragment extends BaseFragment {
    private RecyclerView mRecycler;
    private DatabaseReference mChatDatabase, mUserDatabase, mMessageDatabase;
    private LinearLayoutManager mManager;
    private FirebaseRecyclerAdapter<Conversation, ConversationViewHolder> mAdapter;

    public ConversationFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_conversation, container, false);
        mRecycler = rootView.findViewById(R.id.conv_list);
        mRecycler.setHasFixedSize(true);
        final String user_id = getUid();
        DividerItemDecoration itemDecorator = new DividerItemDecoration(Objects.requireNonNull(getContext()), DividerItemDecoration.VERTICAL);
        itemDecorator.setDrawable(Objects.requireNonNull(ContextCompat.getDrawable(getContext(), R.drawable.divider_friends)));
        mRecycler.addItemDecoration(itemDecorator);

        mChatDatabase = FirebaseDatabase.getInstance().getReference().child("chat").child(user_id);
        mChatDatabase.keepSynced(true);
        mUserDatabase = FirebaseDatabase.getInstance().getReference().child("users");
        mMessageDatabase = FirebaseDatabase.getInstance().getReference().child("messages").child(user_id);
        mUserDatabase.keepSynced(true);

        return rootView;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        mManager = new LinearLayoutManager(getActivity());
        mManager.setReverseLayout(true);
        mManager.setStackFromEnd(true);
        mRecycler.setLayoutManager(mManager);

        Query convQuery = mChatDatabase.orderByChild("timestamp");

        FirebaseRecyclerOptions options = new FirebaseRecyclerOptions.Builder<Conversation>()
                .setQuery(convQuery, Conversation.class)
                .build();

        mAdapter = new FirebaseRecyclerAdapter<Conversation, ConversationViewHolder>(options) {
            @Override
            protected void onBindViewHolder(@NonNull final ConversationViewHolder holder, int position, @NonNull final Conversation model) {
                final String list_user_id = getRef(position).getKey();
                Query lastMessage = mMessageDatabase.child(Objects.requireNonNull(list_user_id)).limitToLast(1);

                lastMessage.addChildEventListener(new ChildEventListener() {
                    @Override
                    public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
                        Message message = dataSnapshot.getValue(Message.class);
                        holder.setMessage(Objects.requireNonNull(message), model.seen);
                    }

                    @Override
                    public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                    }

                    @Override
                    public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {

                    }

                    @Override
                    public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {

                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });

                mUserDatabase.child(list_user_id).addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        final String name = dataSnapshot.child("name").getValue(String.class);
                        final String thumbnail = dataSnapshot.child("thumb_image").getValue(String.class);

                        holder.nameView.setText(name);
                        if (dataSnapshot.hasChild("online")) {
                            String userOnline = dataSnapshot.child("online").getValue(String.class);
                            holder.setUserStatus(Objects.requireNonNull(userOnline));
                        }
                        holder.mView.setOnClickListener(v -> {
                            Intent chatIntent = new Intent(getContext(), ChatActivity.class);
                            chatIntent.putExtra("user_id", list_user_id);
                            chatIntent.putExtra("user_name", name);
                            startActivity(chatIntent);
                        });
                        if (!Objects.requireNonNull(thumbnail).equals("default"))
                            Picasso.get().load(thumbnail).networkPolicy(NetworkPolicy.OFFLINE)
                                    .placeholder(R.drawable.default_avatar).into(holder.userAvatar, new Callback() {
                                @Override
                                public void onSuccess() {

                                }

                                @Override
                                public void onError(Exception e) {
                                    Picasso.get().load(thumbnail).placeholder(R.drawable.default_avatar).into(holder.userAvatar);
                                }
                            });
                        else
                            Picasso.get().load(thumbnail).placeholder(R.drawable.default_avatar).into(holder.userAvatar);
                    }

                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
            }

            @NonNull
            @Override
            public ConversationViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                LayoutInflater inflater = LayoutInflater.from(parent.getContext());
                return new ConversationViewHolder(inflater.inflate(R.layout.item_conv, parent, false));
            }
        };
        mRecycler.setAdapter(mAdapter);
    }

    @Override
    public void onStart() {
        super.onStart();
        if (mAdapter != null)
            mAdapter.startListening();
    }

    @Override
    public void onStop() {
        super.onStop();
        if (mAdapter != null)
            mAdapter.stopListening();
    }

}

File: app/src/main/java/com/a60n1/ejashkojme/fragment/ConversationFragment.kt
package com.a60n1.ejashkojme.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.a60n1.ejashkojme.ChatActivity
import com.a60n1.ejashkojme.R
import com.a60n1.ejashkojme.models.Conversation
import com.a60n1.ejashkojme.models.Message
import com.a60n1.ejashkojme.viewholder.ConversationViewHolder
import com.firebase.ui.database.FirebaseRecyclerAdapter
import com.firebase.ui.database.FirebaseRecyclerOptions
import com.google.firebase.database.*
import com.squareup.picasso.Callback
import com.squareup.picasso.NetworkPolicy
import com.squareup.picasso.Picasso

class ConversationFragment : BaseFragment() {
    private var mRecycler: RecyclerView? = null
    private var mChatDatabase: DatabaseReference? = null
    private var mUserDatabase: DatabaseReference? = null
    private var mMessageDatabase: DatabaseReference? = null
    private var mManager: LinearLayoutManager? = null
    private var mAdapter: FirebaseRecyclerAdapter<Conversation, ConversationViewHolder>? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? { // Inflate the layout for this fragment
        val rootView = inflater.inflate(R.layout.fragment_conversation, container, false)
        mRecycler = rootView.findViewById(R.id.conv_list)
        mRecycler!!.setHasFixedSize(true)
        val userId = uid
        val itemDecorator = DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
        itemDecorator.setDrawable(ContextCompat.getDrawable(context!!, R.drawable.divider_friends)!!)
        mRecycler!!.addItemDecoration(itemDecorator)
        mChatDatabase = FirebaseDatabase.getInstance().reference.child("chat").child(userId)
        mChatDatabase!!.keepSynced(true)
        mUserDatabase = FirebaseDatabase.getInstance().reference.child("users")
        mMessageDatabase = FirebaseDatabase.getInstance().reference.child("messages").child(userId)
        mUserDatabase!!.keepSynced(true)
        return rootView
    }

    @Suppress("UNCHECKED_CAST")
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mManager = LinearLayoutManager(activity)
        mManager!!.reverseLayout = true
        mManager!!.stackFromEnd = true
        mRecycler!!.layoutManager = mManager
        val convQuery = mChatDatabase!!.orderByChild("timestamp")
        val options: FirebaseRecyclerOptions<*> = FirebaseRecyclerOptions.Builder<Conversation>()
                .setQuery(convQuery, Conversation::class.java)
                .build()
        mAdapter = object : FirebaseRecyclerAdapter<Conversation, ConversationViewHolder>(options as FirebaseRecyclerOptions<Conversation>) {
            override fun onBindViewHolder(holder: ConversationViewHolder, position: Int, model: Conversation) {
                val listUserId = getRef(position).key
                val lastMessage = mMessageDatabase!!.child(listUserId!!).limitToLast(1)
                lastMessage.addChildEventListener(object : ChildEventListener {
                    override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {
                        val message = dataSnapshot.getValue(Message::class.java)!!
                        holder.setMessage(message, model.seen)
                    }

                    override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {}
                    override fun onChildRemoved(dataSnapshot: DataSnapshot) {}
                    override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {}
                    override fun onCancelled(databaseError: DatabaseError) {}
                })
                mUserDatabase!!.child(listUserId).addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        val name = dataSnapshot.child("name").value.toString()
                        val thumbnail = dataSnapshot.child("thumb_image").value.toString()
                        holder.nameView.text = name
                        if (dataSnapshot.hasChild("online")) {
                            val userOnline = dataSnapshot.child("online").value.toString()
                            holder.setUserStatus(userOnline)
                        }
                        holder.mView.setOnClickListener {
                            val chatIntent = Intent(context, ChatActivity::class.java)
                            chatIntent.putExtra("user_id", listUserId)
                            chatIntent.putExtra("user_name", name)
                            startActivity(chatIntent)
                        }
                        if (thumbnail != "default") {
                            Picasso.get().load(thumbnail).networkPolicy(NetworkPolicy.OFFLINE)
                                    .placeholder(R.drawable.default_avatar).into(holder.userAvatar, object : Callback {
                                        override fun onSuccess() {}
                                        override fun onError(e: Exception) {
                                            Picasso.get().load(thumbnail).placeholder(R.drawable.default_avatar).into(holder.userAvatar)
                                        }
                                    })
                        } else {
                            Picasso.get().load(thumbnail).placeholder(R.drawable.default_avatar).into(holder.userAvatar)
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError) {}
                })
            }

            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ConversationViewHolder {
                val inflater = LayoutInflater.from(parent.context)
                return ConversationViewHolder(inflater.inflate(R.layout.item_conv, parent, false))
            }
        }
        mRecycler!!.adapter = mAdapter
    }

    override fun onStart() {
        super.onStart()
        if (mAdapter != null) {
            mAdapter!!.startListening()
        }
    }

    override fun onStop() {
        super.onStop()
        if (mAdapter != null) {
            mAdapter!!.stopListening()
        }
    }
}